hibernate-core/src/main/java/org/hibernate/boot/query/SqlResultSetMappingDescriptor.java

```diff
@@ -368,14 +369,22 @@ public FetchMemento resolve(ResultSetMappingResolutionContext resolutionContext)
 			final RuntimeMetamodels runtimeMetamodels = resolutionContext.getSessionFactory().getRuntimeMetamodels();
 			final EntityMappingType entityMapping = runtimeMetamodels.getEntityMappingType( entityName );
 
-			NavigablePath navigablePath = this.navigablePath.append( propertyPathParts[ 0 ] );
 			ModelPart subPart = entityMapping.findSubPart(
-					propertyPathParts[ 0 ],
+					propertyPathParts[0],
 					null
 			);
+			NavigablePath navigablePath;
+			if ( !subPart.getNavigableRole().getParent().equals( entityMapping.getNavigableRole() )
+					&& subPart.getNavigableRole().getParent().getLocalName().equals( ID_ROLE_NAME ) ) {
+				// The attribute is defined in an ID class, append {id} to navigable path
+				navigablePath = this.navigablePath.append( ID_ROLE_NAME ).append( propertyPathParts[0] );
+			}
+			else {
+				navigablePath = this.navigablePath.append( propertyPathParts[0] );
+			}
```

Comment:
I think this should rather be the following to also create the proper navigable path types:

```suggestion
			NavigablePath parentNavigablePath;
			if ( !subPart.getNavigableRole().getParent().equals( entityMapping.getNavigableRole() )
					&& subPart.getNavigableRole().getParent().getLocalName().equals( ID_ROLE_NAME ) ) {
				// The attribute is defined in an ID class, append {id} to navigable path
				parentNavigablePath = new EntityIdentifierNavigablePath( this.navigablePath, null );
			}
			else {
				parentNavigablePath= this.navigablePath;
			}
			if ( subPart instanceof EntityIdentifierMapping ) {
				navigablePath = new EntityIdentifierNavigablePath( parentNavigablePath, propertyPathParts[0] );
			}
			else {
				navigablePath = parentNavigablePath.append( propertyPathParts[0] );
			}
```

Reply:
Right using `EntityIdentifierNavigablePath` is more correct, applied - thanks for the suggestion.